### PR TITLE
feat: Add batched tensor expansion utilities for O(S) ensemble unroll…

### PR DESCRIPTION
**Pull request by Debadri-das**
## Describe your changes

This PR introduces a standalone batch-expanding PyTorch utility (`expand_ensemble_batch` and `fold_ensemble_batch`) to optimize tensor manipulation for Ensemble Lateral Boundary Conditions (LBCs) and probabilistic forecasting variables.

**Motivation and Context:** Currently, probabilistic processing of $S$ ensemble members often relies on sequential `for` loops. To support the upcoming model refactoring and probabilistic capabilities without introducing $O(S)$ sequential bottlenecks, these functions leverage `torch.repeat_interleave()` and `.view()` to dynamically flatten state tensors `(B, ...) -> (B * S, ...)`. This enables optimized parallelized operations over the GPU without permanently mutating standard AR model dimensions.

Dependencies: None. Built strictly with standard PyTorch tensor primitives.

## Issue Link

 #49 and #62 

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch .
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- [ ] Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

File: tests/test_ensemble_utils.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import torch
+from neural_lam.utils import expand_ensemble_batch, fold_ensemble_batch
+
+def test_batched_ensemble_expansion_and_folding():
```
Comment (Sir-Sloth-The-Lazy) [inline tests/test_ensemble_utils.py:4]: A test with `S == T (e.g., S=3, T=3)` would expose the ambiguity.

**Review by Sir-Sloth-The-Lazy — COMMENTED** [review]
Just a heads up you missed the case when `S == T` . Love the idea !

File: neural_lam/utils.py [hunk 2]
```diff
@@ -620,3 +620,52 @@ def get_integer_time(tdelta) -> tuple[int, str]:
...
+        n_members: The ensemble size (S)
+
+    Returns:
+        Tensor of shape (B * n_members, ...)
+    """
+    B = tensor.shape[0]
+    
+    # Check if tensor already has the ensemble dimension S by some heuristic 
+    # or by strictly passing shapes. Assuming standard (B, S, ...) input for LBCs:
+    if tensor.dim() >= 2 and tensor.shape[1] == n_members:
```
Comment (Sir-Sloth-The-Lazy) [inline neural_lam/utils.py]: This is a silent, dangerous assumption. If a deterministic tensor happens to have `shape[1] == n_members `(e.g., n_members=3 and T=3 timesteps), it will be incorrectly treated as having an ensemble dimension and flattened instead of repeated. There's no reliable way to distinguish `(B, S, T, N, F)` from `(B, T, N, F)` by shape alone when `S == T`.

**Review by Sir-Sloth-The-Lazy — COMMENTED** [review]
is this intentional ?

**Comment by sadamov** [comment]
linking #226 and #332 here, to remember the potential connection. this is one for @joeloskarsson to review.

File: tests/test_ensemble_utils.py (hunk 1 above)
Comment (Debadri-das) [reply tests/test_ensemble_utils.py:4]: Done! I've added the exact test case you suggested to `tests/test_ensemble_utils.py`. This fix handles the ambiguity.

**Comment by Debadri-das** [comment]
> is this intentional ?
No, the code was not intentional at all, it was a flaw in the design, which I fixed now.

File: neural_lam/utils.py (hunk 2 above)
Comment (Debadri-das) [reply neural_lam/utils.py]: **Fix:**
I replaced the implicit shape guessing with an **explicit `has_ensemble_dim: bool = False` flag** in the function. Now:
- Users *must* explicitly declare their intent when calling the function
- If `has_ensemble_dim=True` but the tensor lacks proper dimensionality, we raise a clear, actionable `ValueError`
- The `S == T` case you identified is now handled mathematically correctly via the test I added.

**Test:**
Inside `tests/test_ensemble_utils.py`, I added edge case you mentioned:
```python
S_ambig = 3
init_state_ambig = torch.rand(B, S_ambig, N, F)  # Shape (B, T, N, F) where S== T
expanded_ambig = expand_ensemble_batch(init_state_ambig, n_members=S_ambig, has_ensemble_dim=False)
assert expanded_ambig.shape == (B * S_ambig, S_ambig, N, F)

**Comment by Sir-Sloth-The-Lazy** [comment]
@Debadri-das would you be working on this further ? It would be really great if you do or I can take over if you have other work going on. I have this as a dependency for my work :) Hope to hear from you soon !

File: neural_lam/utils.py [hunk 3]
```diff
@@ -644,3 +644,53 @@ def get_integer_time(tdelta: datetime.timedelta) -> tuple[int, str]:
...
+        if tensor.dim() < 2 or tensor.shape[1] != n_members:
+            raise ValueError(f"Expected ensemble dimension of size {n_members} at index 1, but got shape {tensor.shape}")
+        # Already has ensemble dimension, flatten B and S
+        return tensor.view(B * n_members, *tensor.shape[2:])
+    
+    # No ensemble dimension, repeat interleave so elements stay grouped by batch
+    # e.g., [b1, b2] -> [b1, b1, b2, b2]
+    return tensor.repeat_interleave(n_members, dim=0)
+
+def fold_ensemble_batch(tensor: torch.Tensor, n_members: int) -> torch.Tensor:
```
Comment (joeloskarsson) [inline neural_lam/utils.py:677]: It seems like it would be useful for this to be able to take either the number of members or the batch size (and if both are given, check that their product is size of first dim).

File: neural_lam/utils.py [hunk 4]
```diff
@@ -644,3 +644,53 @@ def get_integer_time(tdelta: datetime.timedelta) -> tuple[int, str]:
...
+
+def expand_ensemble_batch(tensor: torch.Tensor, n_members: int, has_ensemble_dim: bool = False) -> torch.Tensor:
+    """
+    Allows for concurrent ensemble processing by expanding a deterministic batch tensor.
+    
+    The input expands (B,...) -> (B * n_members,...) if it lacks an ensemble dimension 
+    (such as beginning states or circumstances shared by all members).
+    It flattens the input to (B * S,...) if it already has an ensemble
+    dimension (for example, perturbed LBCs with shape (B, S,...)).
+
```
Comment (joeloskarsson) [inline neural_lam/utils.py:656]: I think it would be better to have two separate methods for this, as these are two very different things (repeating the same state vs just a reshape). If there is a lot of repeated code add a supporting helper function to avoid this.

Comment (joeloskarsson) [reply neural_lam/utils.py:656]: Happy to discuss if you see a good argument for this being one function!

File: neural_lam/utils.py
Comment (joeloskarsson) [inline neural_lam/utils.py:1]: The added methods here do something very close to `expand_to_batch`
https://github.com/mllam/neural-lam/blob/a208adade0b0d3c38b6741f05231fbd763acd78c/neural_lam/models/step_predictors/base.py#L86-L103
and it seems suboptimal that these methods are in different places in the codebase. Actually, `expand_to_batch` seems like it could just as well sit in utils.py, and there is no reason why it has to be a class method. I think it would make sense to in this PR also move that to utils, to have all batch-dim reordering/reshaping functions in the same place.

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Thanks, added a couple thoughts. Do also check out @Sir-Sloth-The-Lazy's comment, as he is eager to make use of these methods.

**Comment by sadamov** [comment]
Closing this in favour of #649, which consolidates the probabilistic metrics + ensemble plumbing track for v0.8.0. The `expand_ensemble_batch` / `fold_ensemble_batch` utilities will be folded into the consolidated PR. Thanks for the contribution @Debadri-das, you'll be credited.
